lib/Mojolicious/Plugin/OpenAPI/Cors.pm

```diff
@@ -1,43 +1,136 @@
+  return $c if $c->res->headers->header('Access-Control-Allow-Origin');
+
+  # Then go on to preflight
+  $c->openapi->cors_preflighted($cb);
+  return $c if $c->res->headers->header('Access-Control-Allow-Origin');
+
+  # Invalid request
+  return _render_bad_request($c);
+}
+
+sub _helper_cors_preflighted {
+  my ($c, $cb) = @_;
+
+  # Run default simple CORS checks
+  my $method = uc $c->req->method;
+  return $c unless $method eq 'OPTIONS';
+
+  my $req_headers = $c->req->headers;
+  my $req         = {
+    headers => $req_headers->header('Access-Control-Request-Headers') // '',
+    method  => $req_headers->header('Access-Control-Request-Method') // $method,
+    origin  => $req_headers->header('Origin') // '',
+  };
+
+  $req->{type} = $req->{origin} ? 'preflighted' : '';
+
+  # Allow the callback to make up the decision if this is a valid CORS request
+  $c->$cb($req);
+
+  # Valid CORS request if the callback set the Access-Control-Allow-Origin header
+  return _render_preflighted_response($c, $req)
+    if $c->res->headers->header('Access-Control-Allow-Origin');
+
+  # Regular OPTIONS request
+  return $c if $c->stash('openapi_allow_options_request');
+
+  # Invalid if no header is set
+  return $SKIP_RENDER ? $c : _render_bad_request($c);
 }
 
 sub _helper_cors_simple {
   my ($c, $cb) = @_;
-  my $req = {};
+  my $req = {type => 'simple'};
 
   # Run default simple CORS checks
   my $method = uc $c->req->method;
   return $c unless grep { $method eq $_ } @CORS_SIMPLE_METHODS;
 
   my $req_headers = $c->req->headers;
   my $ct = $req_headers->content_type || '';
-  return $c unless grep { $ct eq $_ } @CORS_SIMPLE_CONTENT_TYPES;
+  return $c if $ct and !grep { $ct eq $_ } @CORS_SIMPLE_CONTENT_TYPES;
   return $c unless $req->{origin} = $req_headers->header('Origin');
 
   # Allow the callback to make up the decision if this is a valid CORS request
-  $c->tap($cb, $req);
+  $c->$cb($req);
 
   # Valid CORS request if the callback set the Access-Control-Allow-Origin header
   return $c if $c->res->headers->header('Access-Control-Allow-Origin');
 
   # Invalid if no header is set
-  my $self = $c->stash('openapi.object') or return;
+  return $SKIP_RENDER ? $c : _render_bad_request($c);
+}
+
+sub _render_bad_request {
+  my $c      = shift;
+  my $self   = $c->stash('openapi.object') or return;
   my @errors = ({message => 'Invalid CORS request.'});
   $self->_log($c, '<<<', \@errors);
   $c->render(data => $self->_renderer->($c, {errors => \@errors, status => 400}), status => 400);
   return $c;
 }
 
+sub _render_preflighted_response {
+  my ($c, $req) = @_;
+  my $h = $c->res->headers;
+
+  $h->header('Access-Control-Allow-Headers' => $req->{headers})
+    unless $h->header('Access-Control-Allow-Headers');
+  $h->header('Access-Control-Allow-Methods' => $req->{method})
+    unless $h->header('Access-Control-Allow-Methods');
+  $h->header('Access-Control-Max-Age' => $req->{age} || '3600')
+    unless $h->header('Access-Control-Max-Age');
```

Comment:
I forgot to ask about the default value here. I was considering 600 and 86400 as well, but I think 86400 is way too high for a default.

Maybe you want it to be configurable?